### PR TITLE
Prevent caching normalization results with a cycle

### LINKED ISSUE #79714: rustc hangs and leaks memory
In stable, `rustc 1.48.0 (7eac88abb 2020-11-16)`, `rustc demo.rs` promptly returns `overflow evaluating the requirement` and exits. This is expected behavior.

In nightly, `rustc 1.50.0-nightly (5be3f9f10 2020-12-03)`, `rustc demo.rs` hangs indefinitely, using 100% of one core. Memory usage fluctuates but tends to grow over time at ~800 MB per minute on my machine.

### Source 

```rust
pub trait Baz {}
impl Baz for () {}
impl<T> Baz for [T; 3] {}

pub trait Fiz {}
impl Fiz for bool {}

pub trait Corge<S> {
    type Foo;
}

impl<W: Corge<()>> Corge<[(); 3]> for W {
    type Foo = [<W as Corge<()>>::Foo; 3];
}

pub trait Grault
where
    Self::B: Fiz,
{
    type A;
    type B;
}

impl<T: Grault> Grault for [T; 3]
where
    Self::A: Baz,
    Self::B: Fiz,
{
    type A = [T::A; 3];
    type B = T::B;
}

pub trait Quaks<S, W> {
    fn bar() {}
}

impl<S> Quaks<S, bool> for u8
where
    bool: Corge<S>,
    <bool as Corge<S>>::Foo: Grault,
{
}

fn main() {
    <u8 as Quaks<[(); 3], bool>>::bar();
}
```

`rustc --version --verbose`:
```
rustc 1.50.0-nightly (5be3f9f10 2020-12-03)
binary: rustc
commit-hash: 5be3f9f10e9fd59ea03816840a6051413fbdefae
commit-date: 2020-12-03
host: x86_64-unknown-linux-gnu
release: 1.50.0-nightly
```

### PR DESCRIPTION
**Pull request by matthewjasper**
When normalizing a projection which results in a cycle, we would cache the result of `project_type` without the nested obligations (because they're not needed for inference). This would result in the nested obligations only being handled once in fulfill, which would avoid the cycle error. `get_paranoid_cache_value_obligation` used to add an obligation that resulted in a cycle in this case previously, but was removed by #73905.

This PR makes the projection cache not cache the value of a projection if it was ever normalized in a cycle (except in a snapshot that's rolled back).

Fixes #79714.

r? @nikomatsakis

### CODE REVIEW

File: compiler/rustc_infer/src/traits/project.rs [hunk 1]
```diff
@@ -90,6 +90,7 @@ impl ProjectionCacheKey<'tcx> {
 pub enum ProjectionCacheEntry<'tcx> {
     InProgress,
     Ambiguous,
+    Recur,
```
Comment (bjorn3) [inline compiler/rustc_infer/src/traits/project.rs:93]:
```suggestion
    Recursive,
```
I had to read the doc comment on the `recur` method to understand what this meant.

File: compiler/rustc_infer/src/traits/project.rs [hunk 2]
```diff
@@ -197,6 +203,14 @@ impl<'tcx> ProjectionCache<'_, 'tcx> {
         assert!(!fresh, "never started projecting `{:?}`", key);
     }
 
+    /// Indicates that while trying to normalize `key`, `key` was required to
+    /// be normalized again. Selection or evaluation should eventually report
+    /// an error here.
+    pub fn recur(&mut self, key: ProjectionCacheKey<'tcx>) {
```
Comment (bjorn3) [inline compiler/rustc_infer/src/traits/project.rs:209]:
```suggestion
    pub fn recursive(&mut self, key: ProjectionCacheKey<'tcx>) {
```

**Comment by Mark-Simulacrum** [comment]
@bors r+ p=5 rollup=never

I would like this to get some testing on master before we promote it to beta (and in short order, beta to stable). The patch has received some amount of eyes on it.

**Comment by Mark-Simulacrum** [comment]
Backported to 1.49 in #80417, but leaving accepted for beta backport to 1.50 (this landed after that branched).

**Pull request merged**